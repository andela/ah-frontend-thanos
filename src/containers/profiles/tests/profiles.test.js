import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import EditProfileConnected, { EditProfilePage } from '../editProfile';
import EditProfile from '../../../components/Profile/editProfile';
import ProfileConnected, { Profile } from '../profiles';
import initialStateSample from '../../../commons/initialStates';

Enzyme.configure({ adapter: new Adapter() });
const storeFake = state => ({
  default: jest.fn(),
  subscribe: jest.fn(),
  dispatch: jest.fn(),
  getState: () => state,
});


describe('profile container', () => {
  let wrapper;
  let component;
  let container;

  beforeEach(() => {
    jest.resetAllMocks();
    const profile = {
      username: 'Bruce',
      image: 'htpp://127.0.0.1/test/.png',
      bio: 'This is my bio',
      firstname: 'Bruce',
      lastname: 'Allan',
      isLoggedIn: true,
    };
    const store = storeFake({ profile });

    wrapper = mount(
      <Provider store={store}>
        <EditProfilePage profile={profile} />
      </Provider>,
    );

    container = wrapper.find(EditProfilePage);
    component = wrapper.find(EditProfile);
  });

  it('should render both the container and the component ', () => {
    expect(container.length).toBeTruthy();
    expect(component.length).toBeTruthy();
  });
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Test Profile container', () => {
  let props;
  let initialState;
  let mockStore;

  beforeEach(() => {
    props = {
      match: { params: { username: 'testUser' } },
      getProfileDispatch: jest.fn(),
    };
    initialState = { profileReducer: { profile: { isLoggedIn: true } } };
    mockStore = configureStore([thunk]);
  });

  it('renders the container', () => {
    const { match } = props;
    const wrappers = mount(
      <ProfileConnected
        store={mockStore(initialStateSample)}
        match={match}
        followersList={[]}
        followeesList={[]}
      />,
    );
    expect(wrappers).toMatchSnapshot();
  });
  it('renders the container', () => {
    const { match } = props;
    const wrappers = mount(<EditProfileConnected store={mockStore(initialState)} match={match} />);
    expect(wrappers).toMatchSnapshot();
  });
});

describe('Profile component', () => {
  let wrapper;
  const getProfileAction = jest.fn();
  const nextProps = {
    isLoggedIn: true,
    profilePayload: { followers: ['user2'], following: ['user2'] },
  };
  const props = {
    history: { push: jest.fn() },
  };

  beforeEach(() => {
    const dispatch = jest.fn();
    wrapper = shallow(
      <Profile
        getProfileAction={getProfileAction}
        dispatch={dispatch}
        followersList={[]}
        followeesList={[]}
        {...props}
      />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  // it('should call componentDidMount on rendering', () => {
  //   expect(getProfileAction).toBeCalled();
  // });

  it('should not redirect if isLoggedIn is false', () => {
    wrapper.setProps({ ...nextProps });
    expect(props.history.push).toBeCalledTimes(0);
  });

  it('should redirect to login page if user is not authenticated', () => {
    wrapper.setProps({ isLoggedIn: false });
    expect(props.history.push).toBeCalledWith('/login');
  });
});
