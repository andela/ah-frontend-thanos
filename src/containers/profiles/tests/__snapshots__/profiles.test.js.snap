// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component should render correctly 1`] = `""`;

exports[`Test Profile container renders the container 1`] = `
<Connect(Profile)
  followeesList={Array []}
  followersList={Array []}
  match={
    Object {
      "params": Object {
        "username": "testUser",
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Profile
    dispatch={[Function]}
    followeesList={Array []}
    followersList={Array []}
    history={Object {}}
    isLoggedIn={true}
    match={
      Object {
        "params": Object {
          "username": "testUser",
        },
      }
    }
    profile={
      Object {
        "bio": "",
        "first_name": "",
        "image": "",
        "isLoggedIn": false,
        "last_name": "",
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />
</Connect(Profile)>
`;

exports[`Test Profile container renders the container 2`] = `
<Connect(EditProfilePage)
  match={
    Object {
      "params": Object {
        "username": "testUser",
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <EditProfilePage
    editProfileDispatch={[Function]}
    getUserInputs={[Function]}
    match={
      Object {
        "params": Object {
          "username": "testUser",
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    userProfile={
      Object {
        "isLoggedIn": true,
      }
    }
  >
    <EditProfile
      onChange={[Function]}
      onClick={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="container"
      >
        <h1
          className="title"
        >
          Edit Profile
        </h1>
        <hr />
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            <div
              className="text-center"
            >
              <img
                alt="avatar"
                className="img-fluid"
                src="http://www.macmillanenglish.com/img/author-image.png"
              />
              <h6>
                Upload a different photo...
              </h6>
              <div
                className="col-lg-8 mt-2"
              >
                <button
                  className="btn button"
                  onClick={[Function]}
                  type="submit"
                >
                  Upload
                </button>
              </div>
            </div>
          </div>
          <div
            className="col-md-9 personal-info"
          >
            <h3
              className="pl-3"
            >
              Personal info
            </h3>
            <form
              className="form-horizontal"
              onSubmit={[Function]}
            >
              <FormInput
                el={
                  Object {
                    "elementId": "first_name",
                    "inputType": "text",
                    "placeholder": "First Name",
                  }
                }
                key="first_name"
                onChange={[Function]}
                onClick={[Function]}
              >
                <div
                  className="form-group"
                >
                  <input
                    className="form-control"
                    id="first_name"
                    name="first_name"
                    onChange={[Function]}
                    onClick={[Function]}
                    placeholder="First Name"
                    type="text"
                  />
                </div>
              </FormInput>
              <FormInput
                el={
                  Object {
                    "elementId": "last_name",
                    "inputType": "text",
                    "placeholder": "Last Name",
                  }
                }
                key="last_name"
                onChange={[Function]}
                onClick={[Function]}
              >
                <div
                  className="form-group"
                >
                  <input
                    className="form-control"
                    id="last_name"
                    name="last_name"
                    onChange={[Function]}
                    onClick={[Function]}
                    placeholder="Last Name"
                    type="text"
                  />
                </div>
              </FormInput>
              <textarea
                className="form-control"
                id="exampleFormControlTextarea1"
                name="bio"
                onChange={[Function]}
                placeholder="Tell us something about yourself."
                rows="3"
              />
              <div
                className="col-lg-8"
              >
                <button
                  className="btn  btn-outline-primary mt-4 button"
                  type="submit"
                >
                  Save Changes
                </button>
                <span
                  className="mr-2"
                />
                <button
                  className="btn btn-outline-danger mt-4"
                  type="submit"
                >
                  Cancel
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </EditProfile>
  </EditProfilePage>
</Connect(EditProfilePage)>
`;

exports[`profile container should render correctly 1`] = `
<Provider
  store={
    Object {
      "default": [MockFunction],
      "dispatch": [MockFunction],
      "getState": [Function],
      "subscribe": [MockFunction],
    }
  }
>
  <EditProfilePage
    editProfileDispatch={[Function]}
    getUserInputs={[Function]}
    profile={
      Object {
        "bio": "This is my bio",
        "firstname": "Bruce",
        "image": "htpp://127.0.0.1/test/.png",
        "isLoggedIn": true,
        "lastname": "Allan",
        "username": "Bruce",
      }
    }
    userProfile={Object {}}
  >
    <EditProfile
      onChange={[Function]}
      onClick={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="container"
      >
        <h1
          className="title"
        >
          Edit Profile
        </h1>
        <hr />
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            <div
              className="text-center"
            >
              <img
                alt="avatar"
                className="img-fluid"
                src="http://www.macmillanenglish.com/img/author-image.png"
              />
              <h6>
                Upload a different photo...
              </h6>
              <div
                className="col-lg-8 mt-2"
              >
                <button
                  className="btn button"
                  onClick={[Function]}
                  type="submit"
                >
                  Upload
                </button>
              </div>
            </div>
          </div>
          <div
            className="col-md-9 personal-info"
          >
            <h3
              className="pl-3"
            >
              Personal info
            </h3>
            <form
              className="form-horizontal"
              onSubmit={[Function]}
            >
              <FormInput
                el={
                  Object {
                    "elementId": "first_name",
                    "inputType": "text",
                    "placeholder": "First Name",
                  }
                }
                key="first_name"
                onChange={[Function]}
                onClick={[Function]}
              >
                <div
                  className="form-group"
                >
                  <input
                    className="form-control"
                    id="first_name"
                    name="first_name"
                    onChange={[Function]}
                    onClick={[Function]}
                    placeholder="First Name"
                    type="text"
                  />
                </div>
              </FormInput>
              <FormInput
                el={
                  Object {
                    "elementId": "last_name",
                    "inputType": "text",
                    "placeholder": "Last Name",
                  }
                }
                key="last_name"
                onChange={[Function]}
                onClick={[Function]}
              >
                <div
                  className="form-group"
                >
                  <input
                    className="form-control"
                    id="last_name"
                    name="last_name"
                    onChange={[Function]}
                    onClick={[Function]}
                    placeholder="Last Name"
                    type="text"
                  />
                </div>
              </FormInput>
              <textarea
                className="form-control"
                id="exampleFormControlTextarea1"
                name="bio"
                onChange={[Function]}
                placeholder="Tell us something about yourself."
                rows="3"
              />
              <div
                className="col-lg-8"
              >
                <button
                  className="btn  btn-outline-primary mt-4 button"
                  type="submit"
                >
                  Save Changes
                </button>
                <span
                  className="mr-2"
                />
                <button
                  className="btn btn-outline-danger mt-4"
                  type="submit"
                >
                  Cancel
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </EditProfile>
  </EditProfilePage>
</Provider>
`;
